### PR TITLE
[Quantization] AutoGPTQ refactor and matmul combination support

### PR DESCRIPTION
**Pull request by LeiWang1999**
This PR refactors the AutoGPTQ integration to better align with the framework design. The PR, meanwhile, supports the AutoGPTQ quantization in MLC LLM with matmul combination.

With this PR, you will be able to compile Llama2 using the following command:
```python
python -m mlc_llm.build --model=Llama-2-7b-chat-hf --quantization autogptq_llama_q4f16_1 --target cuda
```
to use the AutoGPTQ quantization. **Note that the first run may take around 10 min for AutoGPTQ quantization computation, and the following runs will be much quicker.** The AutoGPTQ quantization requires the Python `auto_gptq` package to have version at least 0.2.0.

Co-authored-by: Ruihang Lai <ruihangl@cs.cmu.edu>

### CODE REVIEW

**Comment by LeiWang1999** [comment]
Please cc @MasterJH5574

**Comment by LeiWang1999** [comment]
some other dicussions about cross-bytes int3 access.

currently auto-gptq spec in param manager only supports regular lowbit format, like "int2", "int4". Challenges arises when attempting to support iregular formats like "int3". Unlike the  the solution provided by build-in group quant in mlc-llm, which compresses five 3bit into one 16bit, the gptq compresses 32 consecutive 3bit elements into three 32bit int32. this approach sometimes necessitates cross-bit access. I've implemented some tir utils decode func like:

```python
def _tir_offset_u32_to_int_to_float(nbit: int, val: tir.PrimExpr, pos: tir.PrimExpr, offset: tir.PrimExpr, dtype: str):
    assert val.dtype == "uint32"
    mask = tvm.tir.const((1 << nbit) - 1, "uint32")
    return tir.Cast(dtype, (val >> ((pos * nbit) + offset).astype("uint32")) & mask)


def _tir_2xu32_to_int_to_float(nbit: int, val0: tir.PrimExpr, val1: tir.PrimExpr, nb_in_0: tir.PrimExpr, dtype: str):
    assert val0.dtype == "uint32" and val1.dtype == "uint32"
    mask_val0 = (1 << nb_in_0) - 1
    mask_val1 = (1 << nb_in_0)
    mask = tvm.tir.const((1 << nbit) - 1, "uint32")
    offset0 = 32 - nb_in_0
    offset1 = nb_in_0
    val_0 = (((val0 >> offset0).astype("uint32")) & mask_val0).astype("uint32")
    val_1 = (((val1 << offset1).astype("uint32")) & mask_val1).astype("uint32")
    return tir.Cast(dtype, (val_0 | val_1).astype("uint32") & mask)


data_float = tir.Select(
                    nb_in_0 > nbit,
                    tir_utils._tir_offset_u32_to_int_to_float(
                        nbit,
                        qweight[group_idx, j],
                        i % n_float_per_u32,
                        offset,
                        dtype=dtype,
                    ),
                    tir_utils._tir_2xu32_to_int_to_float(
                        nbit,
                        qweight[group_idx, j],
                        qweight[group_idx + 1, j],
                        nb_in_0,
                        dtype=dtype,
                    ),
                )
```

but it will failed when applying the dlight default schedule:

```bash
Traceback (most recent call last):
  File "./build.py", line 4, in <module>
    main()
  File "/workspace/v-leiwang3/mlc-llm/mlc_llm/build.py", line 10, in main
    core.build_model_from_args(parsed_args)
  File "/workspace/v-leiwang3/mlc-llm/mlc_llm/core.py", line 506, in build_model_from_args
    build(mod, args)
  File "/workspace/v-leiwang3/mlc-llm/mlc_llm/core.py", line 418, in build
    mod_deploy = dl.ApplyDefaultSchedule(dl.gpu.GEMV())(mod_deploy)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/ir/transform.py", line 238, in __call__
    return _ffi_transform_api.RunPass(self, mod)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/_ffi/_ctypes/packed_func.py", line 238, in __call__
    raise get_last_ffi_error()
tvm._ffi.base.TVMError: Traceback (most recent call last):
  5: TVMFuncCall
  4: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<tvm::runtime::TypedPackedFunc<tvm::IRModule (tvm::transform::Pass, tvm::IRModule)>::AssignTypedLambda<tvm::transform::{lambda(tvm::transform::Pass, tvm::IRModule)#7}>(tvm::transform::{lambda(tvm::transform::Pass, tvm::IRModule)#7}, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)::{lambda(tvm::runtime::TVMArgs const&, tvm::runtime::TVMRetValue*)#1}> >::Call(tvm::runtime::PackedFuncObj const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, tvm::runtime::TVMRetValue)
  3: tvm::transform::Pass::operator()(tvm::IRModule) const
  2: tvm::transform::Pass::operator()(tvm::IRModule, tvm::transform::PassContext const&) const
  1: tvm::transform::ModulePassNode::operator()(tvm::IRModule, tvm::transform::PassContext const&) const
  0: tvm::runtime::PackedFuncObj::Extractor<tvm::runtime::PackedFuncSubObj<TVMFuncCreateFromCFunc::{lambda(tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*)#2}> >::Call(tvm::runtime::PackedFuncObj const*, tvm::runtime::TVMArgs, tvm::runtime::TVMRetValue*) [clone .cold]
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/_ffi/_ctypes/packed_func.py", line 82, in cfun
    rv = local_pyfunc(*pyargs)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/ir/transform.py", line 307, in _pass_func
    return inst.transform_module(mod, ctx)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/dlight/base/transform.py", line 64, in transform_module
    sch = _apply_rules(func, target, self.rules, tunable=False)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/dlight/base/transform.py", line 80, in _apply_rules
    space = rule.apply(func, target, tunable)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/dlight/gpu/gemv.py", line 173, in apply
    vector_input_buffers = is_gemv(sch, block_info)
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/dlight/gpu/gemv.py", line 88, in is_gemv
    ret = [
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/dlight/gpu/gemv.py", line 91, in <listcomp>
    if len(collect_vars_used_in_access_region(read.region)) < iter_num
  File "/workspace/v-leiwang3/mlc_workspace/tvm-unity/python/tvm/dlight/base/analysis.py", line 220, in collect_vars_used_in_access_region
    assert expr.extent == 1
```

I'll keep tracking the issue if I have time.

File: mlc_llm/core.py
Comment (MasterJH5574) [inline mlc_llm/core.py:486]: Commented out since the GeMV rule fails on AutoGPTQ-quantizated decode-GeMV.

**Please do not merge.**

Comment (MasterJH5574) [reply mlc_llm/core.py]: Good to merge as long as https://github.com/mlc-ai/relax/pull/284 is merged.

**Comment by LeiWang1999** [comment]
LGTM, thanks for your hard work on this pr @MasterJH5574 !

**Review by MasterJH5574 — APPROVED** [review]

**Comment by Lurrobert** [comment]
not supported for macos, unfortunately
https://github.com/PanQiWei/AutoGPTQ/issues/299

do you also think this should be added in the requirements? auto_gptq model

**Pull request merged**
